crates/evm/traces/src/folded_stack_trace.rs

```diff
@@ -0,0 +1,301 @@
+            let contract_name = node.trace.decoded.label.as_ref().unwrap_or(&default_contract_name);
+            format!("new {contract_name}")
+        } else {
+            let selector = node
+                .selector()
+                .map(|selector| selector.encode_hex_with_prefix())
+                .unwrap_or("fallback".to_string());
+            let signature =
+                node.trace.decoded.call_data.as_ref().map(|dc| &dc.signature).unwrap_or(&selector);
+
+            if let Some(label) = &node.trace.decoded.label {
+                format!("{label}.{signature}")
+            } else {
+                signature.clone()
+            }
+        };
+
+        self.fst.enter(func_name, node.trace.gas_used as i64);
+
+        // Track internal function step exits to do in this call context.
+        let mut step_exits = vec![];
+
+        // Process children nodes.
+        for order in &node.ordering {
+            match order {
+                TraceMemberOrder::Call(child_idx) => {
+                    let child_node_idx = node.children[*child_idx];
+                    self.process_call_node(nodes, child_node_idx);
+                }
+                TraceMemberOrder::Step(step_idx) => {
+                    self.exit_previous_steps(&mut step_exits, *step_idx);
+                    self.process_step(&node.trace.steps, *step_idx, &mut step_exits)
+                }
+                TraceMemberOrder::Log(_) => {}
+            }
+        }
+
+        // Exit pending internal function calls if any.
+        for _ in 0..step_exits.len() {
+            self.fst.exit();
+        }
+
+        // Exit from this call context in the folded stack trace.
+        self.fst.exit();
+    }
+
+    /// Creates an entry for an internal function call in the folded stack trace. This method only
+    /// enters the function in the folded stack trace, we cannot exit since we need to exit at a
+    /// future step. Hence, we keep track of the step end index in the `step_exits`.
+    fn process_step(
+        &mut self,
+        steps: &[CallTraceStep],
+        step_idx: usize,
+        step_exits: &mut Vec<usize>,
+    ) {
+        let step = &steps[step_idx];
+        if let Some(decoded_step) = &step.decoded {
+            match decoded_step {
+                DecodedTraceStep::InternalCall(decoded_internal_call, step_end_idx) => {
+                    let gas_used = steps[*step_end_idx].gas_used.saturating_sub(step.gas_used);
+                    self.fst.enter(decoded_internal_call.func_name.clone(), gas_used as i64);
+                    step_exits.push(*step_end_idx);
+                }
+                DecodedTraceStep::Line(_) => {}
+            }
+        }
+    }
+
+    /// Exits all the previous internal calls that should end before starting step_idx.
+    fn exit_previous_steps(&mut self, step_exits: &mut Vec<usize>, step_idx: usize) {
+        let initial_length = step_exits.len();
+        step_exits.retain(|&number| number >= step_idx);
```

Comment:
```suggestion
        step_exits.retain(|&number| number > step_idx);
```

was able to hit this edge case locally (call started right after the terminating step for internal call)